src/controllers/ScrollRestorationController.ts

```diff
@@ -0,0 +1,82 @@
+import { Assert } from '../misc/Assert';
+import { ResultListEvents, IDisplayedNewResultsEventArgs } from '../events/ResultListEvents';
+import { QueryStateModel } from '../models/QueryStateModel';
+import { $$ } from '../utils/Dom';
+import { Logger } from '../misc/Logger';
+
+/**
+ * This class is instantiated on the root of {@link SearchInterface}
+ * if the {@link SearchInterface.options.enableScrollRestoration} option is set to true.
+ *
+ * Its job is to wait until the search results are displayed and the page height is long enough to
+ * restore to the last scroll position. It'll wait for up to `restorationTimeOutInMs` milliseconds for the page to reach its height.
+ *
+ * This is useful when back and forth navigation between the search interface and results is needed on the same tab.
+ *
+ */
+
+export class ScrollRestorationController {
+  public static stateAttributeName: string = 'yoffset';
+  /**
+   * Specifies the max amount of time to wait to restore the last scroll position.
+   */
+  public restorationTimeOutInMs: number = 5000;
+
+  private defaultYOffset: number = 0;
+  private tryToScrollIntervalInMs: number = 50;
+  private timeoutHandle = null;
+
+  constructor(public element: HTMLElement, public window: Window, public queryStateModel: QueryStateModel) {
+    Assert.exists(this.queryStateModel);
+
+    this.initStateAttribute();
+
+    this.window.addEventListener('beforeunload', () => this.saveScrollPositionInState());
```

Comment:
The `this.window` pattern we have in other classes is some kind of artifact due to how we sometimes need to mock window in UT.

Until we get to the point where we have UT and we see the need to mock window, I would simply reference `window` directly, and remove the constructor parameter.

And ideally, most test we will want to do with this will actually be functional/selenium tests, and I have a gut felling that UT will be of relatively low usefulness to verify that it all works well.